pyacq/core/rpc/server.py

```diff
@@ -18,93 +28,465 @@ class RPCServer(object):
+        ## Return object wrapped in LocalObjectProxy _unless_ its type is in noProxyTypes.
+        for typ in no_proxy_types:
+            if isinstance(obj, typ):
+                return obj
+        return self.get_proxy(obj)
+
+
+class QtRPCServer(RPCServer):
+    """RPCServer that lives in a Qt GUI thread.
+    
+    This server may be used to create and manage QObjects, QWidgets, etc. It
+    uses a separate thread to poll for RPC requests, which are then sent to the
+    Qt event loop using by signal (see QtPollThread).
+    
+    Parameters
+    ----------
+    addr : str
+        ZMQ address to listen on. Default is "tcp://*:*".
+    quit_on_close : bool
+        If True, then call `QApplication.quit()` when the server is closed. 
+    """
+    def __init__(self, addr="tcp://*:*", quit_on_close=True):
+        RPCServer.__init__(self, addr)
+        self.quit_on_close = quit_on_close
+        self.poll_thread = QtPollThread(self)
+        
+    def run_forever(self):
+        name = ('%s.%s.%s' % (log.get_host_name(), log.get_process_name(), 
+                              log.get_thread_name()))
+        logging.info("RPC start server: %s@%s", name, self.address.decode())
+        RPCServer.register_server(self)
+        self.poll_thread.start()
+
+    def process_action(self, action, opts, return_type, caller):
+        # this method is called from the Qt main thread.
+        if action == 'close':
+            if self.quit_on_close:
+                QtGui.QApplication.instance().quit()
+            # can't stop poller thread here--that would prevent the return 
+            # message being sent. In general it should be safe to leave this thread
+            # running anyway.
+            #self.poll_thread.stop()
+        return RPCServer.process_action(self, action, opts, return_type, caller)
 
-    def ping(self):
-        return 'pong'
+    def _final_close(self):
+        # Block for a moment to allow the poller thread to flush any pending
+        # messages. Ideally, we could let the poller thread keep the process
+        # alive until it is done, but then we can end up with zombie processes..
+        time.sleep(0.1)
+
+    def _read_and_process_one(self):
+        raise NotImplementedError('Socket reading is handled by poller thread.')
+
+
+class QtPollThread(QtCore.QThread):
+    """Thread that polls an RPCServer socket and sends incoming messages to the
+    server by Qt signal.
+    
+    This allows the RPC actions to be executed in a Qt GUI thread without using
+    a timer to poll the RPC socket. Responses are sent back to the poller
+    thread by a secondary socket.
+    """
+    new_request = QtCore.Signal(object, object)  # client, msg
+    
+    def __init__(self, server):
+        # Note: QThread behaves like threading.Thread(daemon=True); a running
+        # QThread will not prevent the process from exiting.
+        QtCore.QThread.__init__(self)
+        self.server = server
```

Comment:
Is this safe, in the point of view that a Qt python object must not have pointer to its parents ?